src/base/LruMap.h

```diff
@@ -135,11 +153,29 @@ LruMap<EntryValue, EntryCost>::get(const char *key)
     return NULL;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
 bool
-LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
+LruMap<EntryValue, MemoryUsedByEV>::add(const char *key, EntryValue *t)
 {
-    if (ttl == 0)
+    // map the retention concept into a TTL value
+
+    // do not store / retain any entries
+    if (retention == 0)
+        return false;
+
+    // negative retention means store 'forever'
+    if (retention < 0)
+        return add(key, t, 0xFFFFFFFF);
```

Comment:
> Alex: Then, in the three-parameter add(), use numeric_limits to store the maximum time_t value when the entryRetention parameter is negative (or overflows time_t).

Sigh. That would not work because you do want to support negative and zero TTLs in the three-parameter add(). To tie all these lose ends together, we probably should have three add methods:

1. add(key, value): return `add_(key, value, expirationTime)`
2. addWithTtl() or three-parameter add(): check TTL and, for positive values, return `add_(key, value, expirationTime)`
3. add_() or addExpiring(): a private method that does not deal with TTLs and retention. Its third parameter is the expiration time (supplied by the other two methods).